### PR TITLE
ModPC Update V2

### PR DESCRIPTION
**Pull request by itsmeow**
## About The Pull Request

Attempts to get our modPCs up to date again (a sequel to #7338), in further preparation for the PDA-tabletening (pre-req for #7550, basically)

The most important thing in this is the signaller app, but it's mostly just general improvement that will make using the new tablets more enjoyable.

Adds some additional fixes for NTRC because TG's is still broken somehow after two fixes.

Added a filter check to NTRC per admin request.

If you have no ID inserted, the software downloader will say "Insert ID" instead of "No Access". Also fixes the button being recolored to gray instead of being "disabled"

Ports:
- https://github.com/tgstation/tgstation/pull/54716
- https://github.com/tgstation/tgstation/pull/54730
- https://github.com/tgstation/tgstation/pull/56613
- https://github.com/tgstation/tgstation/pull/56662
- https://github.com/tgstation/tgstation/pull/56787
- https://github.com/tgstation/tgstation/pull/57270
- https://github.com/tgstation/tgstation/pull/58895
- https://github.com/tgstation/tgstation/pull/60544
- https://github.com/tgstation/tgstation/pull/61350
- https://github.com/tgstation/tgstation/pull/62533
- https://github.com/tgstation/tgstation/pull/63986
- https://github.com/tgstation/tgstation/pull/64666
- https://github.com/tgstation/tgstation/pull/65298
- https://github.com/tgstation/tgstation/pull/68318
- https://github.com/tgstation/tgstation/pull/58609 **(NTRC CHANGES ONLY)**
- https://github.com/tgstation/tgstation/pull/63433

## Why It's Good For The Game

Tablets are better and more stable.

## Testing Photographs and Procedure
<details>
<summary>Screenshots&Videos</summary>

**Signaller**
image

**Tablet Receives a signal**
image

**Bot Access (no ID insertedd)**
image

**Bot Access (AA inserted or held)**
image

**Summon command works**
image

**Bot Access (Doctor)**
image

**SM moles**
image

**Atmos Scanner C/K**
image

**NTC Observer Chat**
image

The bolding on `to` was removed after this screenshot.

**Insert ID downloader**

image

</details>

## Changelog
:cl: itsmeow, Timberpoes, cacogen, MTandi, Fhaxaris, vincentiusvin, ArcaneMusic, LemonInTheDark, GuillaumePrata, Cheshify, Jackraxxus, OneAsianTortoise, Profakos, tralezab, jlsnow301
tweak: The title bar of Modular PC programs now contains the program's name instead of the file name.
fix: Fixed a bunch of modular PC runtimes with try_eject.
tweak: Anyone can now download BotKeeper (tablet app). Only bots your current ID can unlock will be listed to control.
fix: Fixed NTOS File Manager renaming/deleting wrong programs when used with external disk.
code: Improved ID console name/job sanitization code, and added feedback for when sanitization fails.
fix: Modular ID console will no longer insert/eject ID card for AI or to telekinesis players.
qol: Supermatter Monitor now shows total moles. It glows red if your SM is going to singloose.
code: Supermatter's defines have been moved into it's own file in the __DEFINES folder.
add: Modular Computers equipped with the new integrated radio card chip can now be used as a remote signaller.
add: Scientists now start with tablets pre-equipped with a signaller chip and the program downloaded.
fix: Fixed a bug preventing the Job Management app from de-prioritising jobs when 5 or more jobs were prioritised.
code: Improved computer hardware code and delete management.
qol: The Atmospheric Scanner tablet program will show the tile's temperature as both degrees C and K.
qol: Modular Devices now have auditory feedback for disassembly, and use balloon alerts for various errors.
fix: Signaller code transmissions on the 145.7 frequency are no longer received by all signallers, even if they are tuned to another frequency.
fix: Fixes tablet not updating appearance immediately after it is repaired.
refactor: Modernized NTChat, now with muting and pinging!
admin: NTChat now has a filter check.
qol: Observers can now see NTRC messages if they have PDA messages enabled.
fix: Brig physician slots can no longer be increased through the Job Management tablet app.
tweak: The Software Downloader will now say "Insert ID" instead of "No Access" if no ID is inserted, and no longer responds to hovering the button.
/:cl:

### CODE REVIEW

File: code/modules/modular_computers/NTNet/NTNRC/conversation.dm [hunk 1]
```diff
@@ -16,61 +23,85 @@
 /datum/ntnet_conversation/Destroy()
 	if(SSnetworks.station_network)
 		SSnetworks.station_network.chat_channels.Remove(src)
-	for(var/datum/computer_file/program/chatclient/chatterbox in clients)
+	for(var/datum/computer_file/program/chatclient/chatterbox in (active_clients | offline_clients | muted_clients))
```
Comment (Penwin0) [inline code/modules/modular_computers/NTNet/NTNRC/conversation.dm]: as()?

File: code/modules/modular_computers/NTNet/NTNRC/conversation.dm [hunk 2]
```diff
@@ -16,61 +23,85 @@
...
+	active_clients.Remove(leaving)
+	add_status_message("[leaving.username] has left the channel.")
 
 	// Channel operator left, pick new operator
-	if(C == operator)
+	if(leaving == operator)
 		operator = null
-		if(clients.len)
-			var/datum/computer_file/program/chatclient/newop = pick(clients)
+		if(active_clients.len)
```
Comment (Penwin0) [inline code/modules/modular_computers/NTNet/NTNRC/conversation.dm]: length() is better

File: code/modules/modular_computers/NTNet/NTNRC/conversation.dm [hunk 3]
```diff
@@ -16,61 +23,85 @@
...
+		return
+	if(muted in muted_clients)
+		muted_clients.Remove(muted)
+		muted.computer.alert_call(muted, "You have been unmuted from [title]!", 'sound/machines/ping.ogg')
+	else
+		muted_clients.Add(muted)
+		muted.computer.alert_call(muted, "You have been muted from [title]!")
+
+/datum/ntnet_conversation/proc/ping_user(datum/computer_file/program/chatclient/pinger, datum/computer_file/program/chatclient/pinged)
+	if(pinger in muted_clients) //oh my god fuck off
```
Comment (Penwin0) [inline code/modules/modular_computers/NTNet/NTNRC/conversation.dm]: Dubious comment

File: code/modules/modular_computers/file_system/programs/ntnrc_client.dm [hunk 4]
```diff
@@ -45,13 +48,22 @@
 			var/message = reject_bad_text(params["message"])
 			if(!message)
 				return
-			if(channel.password && !(src in channel.clients))
+			if(src in channel.muted_clients)
+				return
```
Comment (Penwin0) [inline code/modules/modular_computers/file_system/programs/ntnrc_client.dm:53]: Tell them they are muted

File: code/modules/modular_computers/file_system/programs/ntnrc_client.dm [hunk 5]
```diff
@@ -86,8 +98,7 @@
 		if("PRG_toggleadmin")
 			if(netadmin_mode)
 				netadmin_mode = FALSE
-				if(channel)
-					channel.remove_client(src) // We shouldn't be in channel's user list, but just in case...
+				channel?.add_client(src)
```
Comment (Penwin0) [inline code/modules/modular_computers/file_system/programs/ntnrc_client.dm:102]: Intended change?

File: code/modules/modular_computers/file_system/programs/signaller.dm [hunk 6]
```diff
@@ -0,0 +1,99 @@
+/datum/computer_file/program/signaller
+	filename = "signaller"
+	filedesc = "Remote Signaller"
+	category = PROGRAM_CATEGORY_MISC
+	program_icon_state = "signal"
+	extended_desc = "A small built-in frequency app that sends out signaller signals with the appropriate hardware."
+	size = 2
```
Comment (Penwin0) [inline code/modules/modular_computers/file_system/programs/signaller.dm:7]: Only size ones fit in tablets usually, you are putting these in science ones roundstart, if taken out they won't be able to put it back in.

File: code/modules/modular_computers/hardware/sensor_package.dm [hunk 7]
```diff
@@ -6,3 +6,12 @@
 	w_class = WEIGHT_CLASS_TINY
 	device_type = MC_SENSORS
 	expansion_hw = TRUE
+
+/obj/item/computer_hardware/radio_card
```
Comment (Penwin0) [inline code/modules/modular_computers/hardware/sensor_package.dm:10]: File name is no longer accurate

File: code/modules/modular_computers/NTNet/NTNRC/conversation.dm (hunk 1 above)
Comment (itsmeow) [reply code/modules/modular_computers/NTNet/NTNRC/conversation.dm]: Not sure what this means. I've never used as()

File: code/modules/modular_computers/NTNet/NTNRC/conversation.dm (hunk 3 above)
Comment (itsmeow) [reply code/modules/modular_computers/NTNet/NTNRC/conversation.dm]: im sorry it wasn't me 😔 damn tg ports

File: code/modules/modular_computers/file_system/programs/ntnrc_client.dm (hunk 4 above)
Comment (itsmeow) [reply code/modules/modular_computers/file_system/programs/ntnrc_client.dm:53]: their chat box turns red, but yeah good idea

File: code/modules/modular_computers/file_system/programs/signaller.dm (hunk 6 above)
Comment (itsmeow) [reply code/modules/modular_computers/file_system/programs/signaller.dm:7]: This is a program, the sizes are in GQ. The hardware part works fine and can be inserted

File: code/modules/modular_computers/file_system/programs/ntnrc_client.dm (hunk 5 above)
Comment (itsmeow) [reply code/modules/modular_computers/file_system/programs/ntnrc_client.dm:102]: Yep, this fixes a problem where if you toggled it off it wouldn't put you back in the channel.

File: code/modules/modular_computers/NTNet/NTNRC/conversation.dm (hunk 1 above)
Comment (itsmeow) [reply code/modules/modular_computers/NTNet/NTNRC/conversation.dm]: Got it

File: code/__DEFINES/supermatter.dm [hunk 8]
```diff
@@ -0,0 +1,61 @@
+#define SUPERMATTER_MAXIMUM_ENERGY 1e6
+
+#define PLASMA_HEAT_PENALTY 15     // Higher == Bigger heat and waste penalty from having the crystal surrounded by this gas. Negative numbers reduce penalty.
+#define OXYGEN_HEAT_PENALTY 1
```
Comment (Penwin0) [inline code/__DEFINES/supermatter.dm:4]: Moving these defines over to a seperate file seems counterproductive to me.

Comment (itsmeow) [reply code/__DEFINES/supermatter.dm:4]: They need to be global because they're used elsewhere. Leaving them in the SM file means they're not global.

Comment (Penwin0) [reply code/__DEFINES/supermatter.dm:4]: You are right, one of them is

**Review by Penwin0 — APPROVED** [review]

**Review by ike709 — APPROVED** [review]
YOLO

**Pull request merged**